ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/RecoveringSidecarRetriever.java

```diff
@@ -202,28 +237,48 @@ private void maybeInitiateRecovery(
 
   private RecoveryEntry addRecovery(
       final DataColumnSlotAndIdentifier columnId, final BeaconBlock block) {
-    return recoveryBySlot.compute(
-        columnId.slot(),
-        (slot, existingRecovery) -> {
-          if (existingRecovery != null
-              && !existingRecovery.block.getRoot().equals(block.getRoot())) {
-            // we are recovering obsolete column which is no more on our canonical chain
-            existingRecovery.cancel();
-          }
-          if (existingRecovery == null || existingRecovery.cancelled) {
-            return createNewRecovery(block);
-          } else {
-            return existingRecovery;
-          }
-        });
+    final RecoveryEntry recoveryEntry =
+        recoveryBySlot.computeIfAbsent(columnId.slot(), (slot) -> createNewRecovery(block));
+
+    if (!recoveryEntry.blockRoot.equals(block.getRoot())) {
+      LOG.debug(
+          "Found a new block {} at slot {}, pivoting recovery.",
+          block.getRoot(),
+          recoveryEntry.blockRoot);
+      return recoveryBySlot.computeIfPresent(
+          columnId.slot(),
+          (slot, entry) -> {
+            if (entry.blockRoot.equals(recoveryEntry.blockRoot)) {
+              entry.cancel();
+              return createNewRecovery(block);
+            }
+            return entry;
+          });
+    }
+    return recoveryEntry;
```

Comment:
I think the bot is on something.
I don't see why we should nest this.
we could have something like

```java
      recoveryBySlot.compute(columnId.slot(), (slot, existingEntry) -> {
        if (existingEntry != null && existingEntry.blockRoot.equals(block.getRoot())) {
          return existingEntry;
        } else if (existingEntry != null) {
          existingEntry.cancel();
        }
        return createNewRecovery(block);
      });
```
which preserve atomicity